### PR TITLE
Use GCRA for rate limits

### PR DESCRIPTION
**Pull request by kiootic**
ref #3068

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -13,7 +13,7 @@ services:
       - "5432:5432"
 
   redis:
-    image: redis:6.0
+    image: redis:6.2
```
Comment (louischan-oursky) [inline docker-compose.yaml:16]: Does this algorithm require features only available >= 6.2?

Comment (kiootic) [reply docker-compose.yaml:16]: Yes, `PXAT` is added at 6.2

File: pkg/lib/ratelimit/gcra_test.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+			burst:  2,
+			entries: []scheduleEntry{
+				{fn: func() {
+					s.HSet(testKey, "test", "123")
+				}},
+				{time: "0s", n: 0, ok: true, timeToAct: -5},
+				{time: "0s", n: 1, ok: true, timeToAct: 0},
+				{time: "0s", n: -1, ok: true, timeToAct: -5},
+			},
+		})
```
Comment (louischan-oursky) [inline pkg/lib/ratelimit/gcra_test.go:170]: Does the new algorithm handle negative underflow now? Please add a test case to demonstrate it.

Comment (kiootic) [reply pkg/lib/ratelimit/gcra_test.go:170]: It is tested in the reservation test case, let me add some comments to describe the entries.

**Pull request merged**